template/express-rest-api/.buildkite/pipeline.yml

```diff
@@ -30,21 +30,12 @@ configs:
           permit_on_passed: true
 
 steps:
-  - label: 🧖‍♀️ Warm Prod
-    command: ':'
-    key: warm-prod
-    plugins:
-      - *aws-sm
-      - *private-npm
-      - *docker-ecr-cache
```

Comment:
The intent here was to bake the Docker build image and uploading it to ECR once, rather than doing it twice on `Test & Lint` and `Build & Package`. Thoughts?

Reply:
Yeah in my team's repo we call the step `build cache`

Reply:
Yes of course that is still valid.

How about serialising two steps, so it gets done only once - putting the shortest step first (which I think will be `Build & Package`)? Or am I just wrong, and the warm step should remain.

Reply:
If both run at the same time there would be a chance we would get a cache miss right?

Reply:
By "serialise" I meant adding a dependency so that one step always runs after the other, so no chance of a cache miss

Reply:
I think the concern is that a serial approach could work out to be slower overall:

```
Cache 5m
Build 2m | Test 5m = 5m
10m

Build 5m + 2m = 7m
Test 5m
12m
```

It would be great if we had data to substantiate how much we stand to gain/lose in practice. Ideally this would be roughly break even and we can go with the simpler serial approach.

Reply:
Not to mention it might fail while quicker while running both lint and build at the same time.

Reply:
I am investigating / gathering stats.

Reply:
Right, as of the new build queue it looks like each step has to do the pull image step, so the question is whether that pull image step is quicker than running the commands on a single step

Reply:
It would be nice if we could somehow lock the ecr cache plugin to only run once per build at a time, then the first one that runs will build + the others will pull, without the wait step in other scenarios. I am unclear if that is remotely possible though.

Reply:
This definitely helps in UI image, given the UI images are typically alot bigger.

It takes on average 1 min 30 seconds to pull the image on one of our UI repos so moving all the commands to the single step has sped it up a bit